Sporadic failures when building with make -jN

I'm trying to build a Fedora RPM (using their .spec file) from a 4.2.0-based snapshot (i.e., the current HEAD at bfeccae7) to see how it behaves on my machine ahead of the 4.2.0 release.

On some build attempts (initially on the first couple of them, which is why I noticed the problem at all; then later, couple of them succeeded and some failed), the build process fails when trying to build the desktop file.

Example error:

<details><summary>Full build error</summary><p>

```
$ rpmbuild -ba exaile.spec 
Building target platforms: noarch
Building for target noarch
setting SOURCE_DATE_EPOCH=1760745600
Executing(%mkbuilddir): /bin/sh -e /var/tmp/rpm-tmp.yMepOF
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.ysO25P
+ umask 022
+ cd /home/rok/rpmbuild/BUILD/exaile-4.2.0-build
+ cd /home/rok/rpmbuild/BUILD/exaile-4.2.0-build
+ rm -rf exaile-4.2.0
+ /usr/lib/rpm/rpmuncompress -x /home/rok/rpmbuild/SOURCES/exaile-4.2.0.tar.gz
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd exaile-4.2.0
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.Zk5I8f
+ umask 022
+ cd /home/rok/rpmbuild/BUILD/exaile-4.2.0-build
+ CFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer '
+ export CFLAGS
+ CXXFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer '
+ export CXXFLAGS
+ FFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -I/usr/lib/gfortran/modules '
+ export FFLAGS
+ FCFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -I/usr/lib/gfortran/modules '
+ export FCFLAGS
+ VALAFLAGS=-g
+ export VALAFLAGS
+ RUSTFLAGS='-Copt-level=3 -Cdebuginfo=2 -Ccodegen-units=1 -Cstrip=none -Cforce-frame-pointers=yes --cap-lints=warn'
+ export RUSTFLAGS
+ LDFLAGS='-Wl,-z,relro -Wl,--as-needed   -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1  '
+ export LDFLAGS
+ LT_SYS_LIBRARY_PATH=/usr/lib:
+ export LT_SYS_LIBRARY_PATH
+ CC=gcc
+ export CC
+ CXX=g++
+ export CXX
+ cd exaile-4.2.0
+ CFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer '
+ export CFLAGS
+ CXXFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer '
+ export CXXFLAGS
+ FFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -I/usr/lib/gfortran/modules '
+ export FFLAGS
+ FCFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -I/usr/lib/gfortran/modules '
+ export FCFLAGS
+ VALAFLAGS=-g
+ export VALAFLAGS
+ RUSTFLAGS='-Copt-level=3 -Cdebuginfo=2 -Ccodegen-units=1 -Cstrip=none -Cforce-frame-pointers=yes --cap-lints=warn'
+ export RUSTFLAGS
+ LDFLAGS='-Wl,-z,relro -Wl,--as-needed   -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1  '
+ export LDFLAGS
+ LT_SYS_LIBRARY_PATH=/usr/lib:
+ export LT_SYS_LIBRARY_PATH
+ CC=gcc
+ export CC
+ CXX=g++
+ export CXX
+ sed -i 's|install -m|$(INSTALL) -m|;s|all: compile |all: |' Makefile
+ sed -i 's|BAD = \[\]|BAD = ['\''daapclient'\'', '\''daapserver'\'', '\''moodbar'\'', '\''winmmkeys'\'', '\''wikipedia'\'']|' plugins/list.py
+ /usr/bin/make -O -j12 V=1 VERBOSE=1
mkdir -p build
python3 tools/generate-completion.py bash > build/exaile.bash-completion
python3 tools/generate-completion.py fish > build/exaile.fish-completion
LC_ALL=C help2man -n "music manager and player" -N ./exaile \
  | gzip -9 > build/exaile.1.gz
echo af ar ast az be be@latin bg bn bs ca cs csb cy da de el en_AU en_CA en_GB eo es et eu fa fi fo fr frp fy gl gu he hi hr hu id ie it ja ka kk ko lt lv mk ml mr ms nb nl oc os pl pt pt_BR ro ru sc si sk sl sq sr sv sw ta te th tl tr ts uk ur vi zh zh_CN zh_TW > po/LINGUAS
msgfmt --desktop --template=data/exaile.desktop.in -d po -o build/exaile.desktop
po/sc.po:5366: end-of-file within string
po/sc.po:5365: missing 'msgstr' section
msgfmt: found 2 fatal errors
make: *** [Makefile:303: desktop_files] Error 1
make: *** Waiting for unfinished jobs....
( ( set -o pipefail 2> /dev/null ) && set -o pipefail ; \
  export LC_ALL=C && cd po && \
  { find ../xl ../xlgui -name "*.py" | sort && \
    find ../data/ui -name "*.ui" | sort && \
    find ../plugins -name "*.py" | sort && \
    find ../plugins -name "*.ui" | sort ; } \
  | xgettext --files-from=- --output=messages.pot --from-code=UTF-8 --add-comments=TRANSLATORS --keyword=N_ && \
  find ../plugins -name PLUGININFO | sort \
  | xgettext --files-from=- --output=messages.pot --from-code=UTF-8 --add-comments=TRANSLATORS --join-existing --language=Python && \
  xgettext -j -o messages.pot --keyword=X-GNOME-FullName '../data/exaile.desktop.in' && \
  xgettext -j -o messages.pot '../data/exaile.appdata.xml.in' )
../xl/formatter.py:630: warning: 'msgid' format string with unnamed arguments cannot be properly localized:
                                 The translator cannot reorder the arguments.
                                 Please consider using a format string with named arguments,
                                 and a mapping instead of a tuple for the arguments.
../xl/formatter.py:633: warning: 'msgid' format string with unnamed arguments cannot be properly localized:
                                 The translator cannot reorder the arguments.
                                 Please consider using a format string with named arguments,
                                 and a mapping instead of a tuple for the arguments.
../xl/player/gst/missing_plugin.py:88: warning: 'msgid' format string with unnamed arguments cannot be properly localized:
                                                The translator cannot reorder the arguments.
                                                Please consider using a format string with named arguments,
                                                and a mapping instead of a tuple for the arguments.
../xl/player/gst/sink.py:192: warning: 'msgid' format string with unnamed arguments cannot be properly localized:
                                       The translator cannot reorder the arguments.
                                       Please consider using a format string with named arguments,
                                       and a mapping instead of a tuple for the arguments.
../xlgui/widgets/dialogs.py:1681: warning: 'msgid' format string with unnamed arguments cannot be properly localized:
                                           The translator cannot reorder the arguments.
                                           Please consider using a format string with named arguments,
                                           and a mapping instead of a tuple for the arguments.
../xlgui/widgets/dialogs.py:1698: warning: 'msgid' format string with unnamed arguments cannot be properly localized:
                                           The translator cannot reorder the arguments.
                                           Please consider using a format string with named arguments,
                                           and a mapping instead of a tuple for the arguments.
../plugins/bpm/__init__.py:137: warning: 'msgid' format string with unnamed arguments cannot be properly localized:
                                         The translator cannot reorder the arguments.
                                         Please consider using a format string with named arguments,
                                         and a mapping instead of a tuple for the arguments.
../plugins/cd/PLUGININFO:4: warning: 'msgid' format string with unnamed arguments cannot be properly localized:
                                     The translator cannot reorder the arguments.
                                     Please consider using a format string with named arguments,
                                     and a mapping instead of a tuple for the arguments.
../plugins/somafm/PLUGININFO:4: warning: 'msgid' format string with unnamed arguments cannot be properly localized:
                                         The translator cannot reorder the arguments.
                                         Please consider using a format string with named arguments,
                                         and a mapping instead of a tuple for the arguments.
find po -name '*.po' -exec \
  msgmerge --previous --no-fuzzy-matching --update {} po/messages.pot \;
................................................................................................................................................................................... done.
........................................................................................................................................................................... done.
................................................................................................................................................................................ done.
........................................................................................................................................................................................ done.
............................................................................................................................................................................ done.
........................................................................................................................................................................... done.
............................................................................................................................................................................................ done.
.............................................................................................................................................................................................. done.
................................................................................................................................................................................. done.
.............................................................................................................................................................................................. done.
......................................................................................................................................................................... done.
......................................................................................................................................................................... done.
.............................................................................................................................................................................. done.
................................................................................................................................................................................................... done.
.......................................................................................................................................................................... done.
........................................................................................................................................................................... done.
............................................................................................................................................................................. done.
.................................................................................................................................................................................. done.
........................................................................................................................................................................................ done.
.................................................................................................................................................................. done.
.......................................................................................................................................................................................... done.
....................................................................................................................................................................... done.
..................................................................................................................................................................................... done.
................................................................................................................................................................................... done.
............................................................................................................................................................ done.
................................................................................................................................................................................... done.
........................................................................................................................................................................................... done.
....................................................................................................................................................................... done.
............................................................................................................................................................................... done.
...................................................................................................................................................................................... done.
.............................................................................................................................................................................. done.
...................................................................................................................................................................................... done.
.............................................................................................................................................................................. done.
..................................................................................................................................................................... done.
................................................................................................................................................................... done.
......................................................................................................................................................................................... done.
............................................................................................................................................................................. done.
................................................................................................................................................................ done.
.......................................................................................................................................................................... done.
........................................................................................................................................................................ done.
........................................................................................................................................................................ done.
................................................................................................................................................................................ done.
.................................................................................................................................................................................. done.
............................................................................................................................................................................... done.
............................................................................................................................................................................... done.
............................................................................................................................................................................... done.
...................................................................................................................................................................................... done.
............................................................................................................................................................................. done.
.................................................................................................................................................................................... done.
.................................................................................................................................................................... done.
................................................................................................................................................................................. done.
........................................................................................................................................................................... done.
................................................................................................................................................................................................ done.
.......................................................................................................................................................................................... done.
..................................................................................................................................................................... done.
............................................................................................................................................................................ done.
........................................................................................................................................................................... done.
...................................................................................................................................................................................... done.
................................................................................................................................................................................ done.
.............................................................................................................................................................................. done.
............................................................................................................................................................................. done.
..................................................................................................................................................................... done.
................................................................................................................................................................ done.
......................................................................................................................................................................... done.
.................................................................................................................................................................................. done.
......................................................................................................................................................................................... done.
.................................................................................................................................................................................. done.
.................................................................................................................................................................................. done.
...................................................................................................................................................................... done.
............................................................................................................................................................................ done.
................................................................................................................................................................................ done.
............................................................................................................................................................................. done.
..................................................................................................................................................................................... done.
................................................................................................................................................................................. done.
............................................................................................................................................................................ done.
................................................................................................................................................................................. done.
................................................................................................................................................. done.
error: Bad exit status from /var/tmp/rpm-tmp.Zk5I8f (%build)

RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.Zk5I8f (%build)
```

</p></details>

Or just the relevant extract from the above:

```
+ /usr/bin/make -O -j12 V=1 VERBOSE=1
mkdir -p build
python3 tools/generate-completion.py bash > build/exaile.bash-completion
python3 tools/generate-completion.py fish > build/exaile.fish-completion
LC_ALL=C help2man -n "music manager and player" -N ./exaile \
  | gzip -9 > build/exaile.1.gz
echo af ar ast az be be@latin bg bn bs ca cs csb cy da de el en_AU en_CA en_GB eo es et eu fa fi fo fr frp fy gl gu he hi hr hu id ie it ja ka kk ko lt lv mk ml mr ms nb nl oc os pl pt pt_BR ro ru sc si sk sl sq sr sv sw ta te th tl tr ts uk ur vi zh zh_CN zh_TW > po/LINGUAS
msgfmt --desktop --template=data/exaile.desktop.in -d po -o build/exaile.desktop
po/sc.po:5366: end-of-file within string
po/sc.po:5365: missing 'msgstr' section
msgfmt: found 2 fatal errors
make: *** [Makefile:303: desktop_files] Error 1
make: *** Waiting for unfinished jobs....
```

Or from a different run:

```
+ /usr/bin/make -O -j12 V=1 VERBOSE=1
mkdir -p build
python3 tools/generate-completion.py bash > build/exaile.bash-completion
python3 tools/generate-completion.py fish > build/exaile.fish-completion
LC_ALL=C help2man -n "music manager and player" -N ./exaile \
  | gzip -9 > build/exaile.1.gz
echo af ar ast az be be@latin bg bn bs ca cs csb cy da de el en_AU en_CA en_GB eo es et eu fa fi fo fr frp fy gl gu he hi hr hu id ie it ja ka kk ko lt lv mk ml mr ms nb nl oc os pl pt pt_BR ro ru sc si sk sl sq sr sv sw ta te th tl tr ts uk ur vi zh zh_CN zh_TW > po/LINGUAS
msgfmt --desktop --template=data/exaile.desktop.in -d po -o build/exaile.desktop
msgfmt --xml --template=data/exaile.appdata.xml.in -d po -o build/org.exaile.exaile.appdata.xml
po/es.po:5421: keyword "msgi" unknown
po/es.po:5421:8: syntax error
msgfmt: found 2 fatal errors
make: *** [Makefile:304: desktop_files] Error 1
make: *** Waiting for unfinished jobs....
```

The reported .po file may change between the runs, and so may the allegedly problematic line.

The .spec uses `%make_build` macro, which expands into `/usr/bin/make -O -j${RPM_BUILD_NCPUS} V=1 VERBOSE=1` that you can see in the above examples.

Also, I noticed that running `make` in the git checkout results in .po files being modified.

So I suspect that with a parallel run, the desktop file generation step ends up running in parallel with building of the locale .mo files; as the latter modifies the .po files, we end up with a race condition where the desktop generation may end up using a partially-written .po file.

I am not sure if this is a new issue or not; perhaps prior to https://github.com/exaile/exaile/commit/004130774beeb0284bca2f6f87eb66d59ddf450f, the .mo files (and hence .po files) were built before desktop files? (Although I'm not sure that old `locale: builddir $(LOCALE_OBJS) desktop_files` should imply anything about the order of dependent targets). But that change was is already is 4.1.4, so I would expect Fedora packagers to run into this as well, at least on some build runs...

--- a/Makefile
+++ b/Makefile
@@ -298,7 +298,7 @@ format:
 check_format:
 	$(BLACK) --check --diff -S *.py plugins/ xl/ xlgui/ tests/ tools/installer/exaile.spec
 
-desktop_files: builddir
+desktop_files: builddir locale
 	echo $(LINGUAS) > po/LINGUAS
 	msgfmt --desktop --template=data/exaile.desktop.in -d po -o build/exaile.desktop
 	msgfmt --xml --template=data/exaile.appdata.xml.in -d po -o build/org.exaile.exaile.appdata.xml

